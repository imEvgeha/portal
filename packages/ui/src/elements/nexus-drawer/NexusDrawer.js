import React from 'react';
import PropTypes from 'prop-types';
import Blanket from '@atlaskit/blanket';
import Page, {Grid, GridColumn} from '@atlaskit/page';
import Spinner from '@atlaskit/spinner';
import {Button} from '@portal/portal-components';
import classnames from 'classnames';
import './NexusDrawer.scss';

const NexusDrawer = ({
    children,
    width,
    isOpen,
    isLoading = true,
    onClose,
    title,
    headerContent,
    isClosedOnBlur,
    direction,
}) => {
    const drawerClassNames = classnames('nexus-c-drawer', {
        'nexus-c-drawer--from-left': direction === 'fromLeft',
        'nexus-c-drawer--from-right': direction === 'fromRight',
        [`nexus-c-drawer--is-${width}-open`]: isOpen,
    });

    const titleContainerClassNames = classnames('nexus-c-drawer__title-container', {
        'nexus-c-drawer__title-container--from-left': direction === 'fromLeft',
    });

    const headerClassNames = classnames('nexus-c-drawer__header', {
        'nexus-c-drawer__header--from-left': direction === 'fromLeft',
        'nexus-c-drawer__header--from-right': direction === 'fromRight',
        'nexus-c-drawer__header--is-open': isOpen,
    });

    return (
        <>
            {isOpen && <Blanket onBlanketClicked={isClosedOnBlur ? onClose : null} canClickThrough={false} isTinted />}
            <div className={drawerClassNames}>
                <Page>
                    {isLoading ? (
                        <div className="nexus-c-drawer--is-loading">
                            <Spinner size="large" />
                        </div>
                    ) : (
                        <>
                            <Grid layout="fluid">
                                <GridColumn>
                                    <div className={headerClassNames}>
                                        <div className={titleContainerClassNames}>
                                            <h1 className="nexus-c-drawer__title">{title}</h1>
                                            {!!headerContent && (
                                                <>
                                                    <div className="nexus-c-drawer__line-break" />
                                                    <div className="nexus-c-drawer__title-content">{headerContent}</div>
                                                </>
                                            )}
                                        </div>
                                        <Button
                                            className="p-button-text"
                                            icon="po po-remove"
                                            tooltip="Close Drawer"
                                            onClick={onClose}
                                        />
                                    </div>
                                </GridColumn>
                            </Grid>
                            {isOpen && children}
                        </>
                    )}
                </Page>
            </div>
        </>
    );
};

NexusDrawer.propTypes = {
    title: PropTypes.string,
    width: PropTypes.string,
    isOpen: PropTypes.bool.isRequired,
    isLoading: PropTypes.bool,
    onClose: PropTypes.func.isRequired,
    headerContent: PropTypes.element,
    isClosedOnBlur: PropTypes.bool,
    direction: PropTypes.oneOf(['fromLeft', 'fromRight']),
};

NexusDrawer.defaultProps = {
    title: '',
    width: 'medium',
    headerContent: null,
    isClosedOnBlur: true,
    isLoading: false,
    direction: 'fromRight',
};

export default NexusDrawer;
