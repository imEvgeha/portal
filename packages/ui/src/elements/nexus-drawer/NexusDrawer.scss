$widths: (
    'extended': 95vw,
    'full': 100vw,
    'medium': 480px,
    'narrow': 360px,
    'wide': 800px,
    'wider': 96.5vw,
);

.nexus-c-drawer {
    position: absolute;
    top: 0;
    bottom: 0;
    overflow: auto;
    width: 0;
    background-color: white;
    z-index: 500;
    transition: width 220ms cubic-bezier(0.2, 0, 0, 1);

    &--from-left {
        left: 0;
    }

    &--from-right {
        right: 0;
    }

    @each $width, $size in $widths {
        &--is-#{$width}-open {
            width: map-get($map: $widths, $key: $width);
        }
    }

    &--is-loading {
        width: 100%;
        height: 100%;
        display: flex;
        justify-content: center;
        align-items: center;
    }

    &__title {
        margin-left: 0 !important;
    }

    &__title-container {
        margin-left: 0;

        &--from-left {
            margin-left: 1rem;
        }
    }

    &__title-content {
        margin-top: 0.5rem;
    }

    &__header {
        margin-top: 1rem;
        margin-bottom: 1rem;
        display: none;

        &--is-open {
            display: flex;
            align-items: flex-start;
            flex-wrap: wrap;
        }

        &--from-left {
            justify-content: flex-end;
            flex-direction: row-reverse;
        }

        &--from-right {
            justify-content: space-between;
        }
    }
}
