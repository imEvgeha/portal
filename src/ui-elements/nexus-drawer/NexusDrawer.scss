@import '../../styles/config';

.nexus-c-drawer {
  position: absolute;
  top: 0;
  height: 100%;
  background-color: red;
  z-index: map-get($app-layout, z-index-above-overlay);
  transition: transform 200ms ease-in;

  &--is-right {
    right: 0;
  }

  &--is-left {
    left: -400px;
  }

  &--is-small-width {
    width: 200px;

    & .nexus-c-drawer--is-open {
      transform: translateX(-200px);
    }
  }

  &--is-medium-width {
    width: 400px;

    & .nexus-c-drawer--is-open {
      transform: translateX(400px);
    }
  }

  &--is-large-width {
    width: 600px;

    & .nexus-c-drawer--is-open {
      transform: translateX(-600px);
    }
  }
}
