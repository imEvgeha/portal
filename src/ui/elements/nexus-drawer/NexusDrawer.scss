@import '../../../styles/config';

$width: (
  extended: 95vw,
  full: 100vw,
  medium: 480px,
  narrow: 360px,
  wide: 600px,
);
$sizes: full, narrow, extended, medium, wide;
$positions: left, right;

.nexus-c-drawer {
  position: absolute;
  top: 0;
  height: 100%;
  background-color: map-get($colors, nexus-white);
  z-index: map-get($z-index, z-index-1000);
  transition: transform 200ms ease-in;

  // Creates classes for each combination of positions and sizes
  @each $position in $positions {
    @each $size in $sizes {
      // Move the drawer off-screen based on position and width
      &--is-#{$position}#{&}--is-#{$size}-width {
        #{$position}: -#{map-get($width, #{$size})};
      }

      // Set drawer widths
      &--is-#{$size}-width {
        width: map-get($width, #{$size});
      }

      // When drawer is open, apply transformation based on position and size
      &--is-#{$size}-width#{&}--is-open {
        &.nexus-c-drawer--is-right {
          // If the drawer is on the right side, we want it moving to the left side
          transform: translateX(-#{map-get($width, #{$size})});
        }
        &.nexus-c-drawer--is-left {
          // If the drawer is on the left side, we want it moving to the right side
          transform: translateX(map-get($width, #{$size}));
        }
      }
    }
  }
}

.nexus-c-drawer__close-btn {
  position: absolute;
  top: 0;

  @each $position in $positions {
    &--is-#{$position} {
      #{$position}: 0;
    }
  }
}
