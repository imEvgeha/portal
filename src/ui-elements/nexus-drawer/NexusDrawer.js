import React, {Fragment, useContext, useEffect, useState} from 'react';
import PropTypes from 'prop-types';
import ArrowLeftIcon from '@atlaskit/icon/glyph/arrow-left';
import ArrowRightIcon from '@atlaskit/icon/glyph/arrow-right';
import {NexusOverlayContext} from '../nexus-overlay/NexusOverlay';
import Button from '@atlaskit/button';

import './NexusDrawer.scss';

const NexusDrawer = ({
    children,
    position,
    width,
    isOpen,
    onClose,
}) => {
    const {setIsOverlayActive} = useContext(NexusOverlayContext);

    const [isOpenClass, setIsOpenClass] = useState('');

    useEffect(() => {
        setIsOverlayActive(isOpen);
        setIsOpenClass(isOpen ? 'nexus-c-drawer--is-open' : '');
        }, [isOpen]
    );

    return (
        <div
            className={`
                nexus-c-drawer 
                nexus-c-drawer--is-${position} 
                nexus-c-drawer--is-${width}-width
                ${isOpenClass} 
            `}
        >
            <Button
                appearance="subtle"
                onClick={onClose}
                className={`nexus-c-drawer__close-btn nexus-c-drawer__close-btn--is-${position}`}
            >
                {position === 'right' ? <ArrowRightIcon /> : <ArrowLeftIcon />}
            </Button>
            {children}
        </div>
    );
};

NexusDrawer.propTypes = {
    position: PropTypes.string,
    width: PropTypes.string,
    isOpen: PropTypes.bool.isRequired,
    onClose: PropTypes.func.isRequired,
};

NexusDrawer.defaultProps = {
    position: 'right',
    width: 'medium',
};

export default NexusDrawer;
