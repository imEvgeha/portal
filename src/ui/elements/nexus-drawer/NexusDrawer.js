import React, {useEffect, useState} from 'react';
import PropTypes from 'prop-types';
import Blanket from '@atlaskit/blanket';
import CloseIcon from '../../../assets/action-cross.svg';
import './NexusDrawer.scss';

const NexusDrawer = ({
    children,
    width,
    isOpen,
    onClose,
    title,
    headerContent,
    isClosedOnBlur,
}) => {
    const [isOpenClass, setIsOpenClass] = useState('');

    useEffect(
        () => setIsOpenClass(isOpen ? 'nexus-c-drawer--is-open' : ''),
        [isOpen]
    );

    return (
        <>
            {isOpen && (
                <Blanket
                    onBlanketClicked={isClosedOnBlur ? onClose : null}
                    canClickThrough={false}
                    isTinted
                />
            )}
            <div
                className={
                    `nexus-c-drawer
                    nexus-c-drawer--is-${width}-width
                    ${isOpenClass}`
                }
            >
                <div className="nexus-c-drawer__header">
                    <div className="nexus-c-drawer__header--title">{title}</div>
                    <CloseIcon className="nexus-c-drawer__header--close-btn" onClick={onClose} />
                    {!!headerContent && (
                        <>
                            <div className="break" />
                            <div className="nexus-c-drawer__header__bottom">
                                {headerContent}
                            </div>
                        </>
                    )}
                </div>
                {isOpen && children}
            </div>
        </>
    );
};

NexusDrawer.propTypes = {
    title: PropTypes.string,
    width: PropTypes.string,
    isOpen: PropTypes.bool.isRequired,
    onClose: PropTypes.func.isRequired,
    headerContent: PropTypes.element,
    isClosedOnBlur: PropTypes.bool,
};

NexusDrawer.defaultProps = {
    title: '',
    width: 'medium',
    headerContent: null,
    isClosedOnBlur: true,
};

export default NexusDrawer;
