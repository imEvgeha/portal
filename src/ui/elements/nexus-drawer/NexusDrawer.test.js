import React from 'react';
import {shallow} from 'enzyme';

import NexusDrawer from './NexusDrawer';

describe('NexusDrawer', () => {
    let wrapper;
    it('should match snapshot', () => {
        wrapper = shallow(
            <NexusDrawer
                position="right"
                width="narrow"
                isOpen={true}
                onClose={() => null}
            />
        );
        expect(wrapper).toMatchSnapshot();
    });

    it('should not render children when closed', () => {
        wrapper = shallow(
            <NexusDrawer
                onClose={() => null}
                isOpen={false}
            >
                <div>Child</div>
            </NexusDrawer>
        );
        expect(wrapper.contains(<div>Child</div>)).toBeFalsy();
    });

    it('should render children when opened', () => {
        wrapper = shallow(
            <NexusDrawer
                onClose={() => null}
                isOpen={true}
            >
                <div>Child</div>
            </NexusDrawer>
        );
        expect(wrapper.contains(<div>Child</div>)).toBeTruthy();
    });

    it('should have only one close button and call onClose when button is clicked', () => {
        const onClose = jest.fn();
        wrapper = shallow(
            <NexusDrawer
                onClose={onClose}
                isOpen={true}
            />
        );

        expect(wrapper.find('.nexus-c-drawer__close-btn')).toHaveLength(1);
        wrapper.find('.nexus-c-drawer__close-btn').simulate('click');
        expect(onClose).toHaveBeenCalled();
    });
});

