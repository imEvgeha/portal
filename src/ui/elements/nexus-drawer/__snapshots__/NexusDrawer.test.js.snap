// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NexusDrawer should match snapshot 1`] = `
<Fragment>
  <WithAnalyticsContext(WithAnalyticsEvents(Blanket))
    canClickThrough={false}
    isTinted={true}
    onBlanketClicked={[Function]}
  />
  <div
    className="nexus-c-drawer 
                    nexus-c-drawer--is-narrow-width
                    "
  >
    <div
      className="nexus-c-drawer__header"
    >
      <div
        className="nexus-c-drawer__header--title"
      />
      <[object Object]
        onClick={[Function]}
      />
    </div>
  </div>
</Fragment>
`;
